File: plugin/suite/roster/orchestration/src/upgrade.py
#!/usr/bin/env python3
"""Check for and apply Cadre CLI updates.

Usage:
    cadre upgrade            # Check for updates and prompt if available
    cadre upgrade --check    # Only check, don't prompt or update
    cadre upgrade --force    # Update without confirmation
    cadre upgrade --help     # Show this help

This tool checks for newer versions available on GitHub and offers to update
the current installation using the package manager that installed Cadre
(detected automatically: pip, pipx, or source checkout).
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from pathlib import Path
from typing import Any

try:
    import urllib.request
    import urllib.error
except ImportError as e:
    print(f"cadre upgrade: requires Python's urllib (standard library): {e}", file=sys.stderr)
    sys.exit(1)

def _find_repo_root() -> Path:
    """Find repository root by searching for cadre_cli/_version.py."""
    current = Path(__file__).resolve().parent
    for _ in range(10):  # Search up to 10 levels
        if (current / "cadre_cli" / "_version.py").is_file():
            return current
        current = current.parent
    # Fallback: assume standard location (3 levels up from src/)
    return Path(__file__).resolve().parents[3]

REPO_ROOT = _find_repo_root()


def get_installed_version() -> str:
    """Get the currently installed Cadre version."""
    import ast

    checkout_marker = REPO_ROOT / "cadre_cli" / "_version.py"
    version_marker = checkout_marker if checkout_marker.is_file() else REPO_ROOT.parent / "_version.py"

    try:
        module = ast.parse(version_marker.read_text(encoding="utf-8"), filename=str(version_marker))
    except (OSError, SyntaxError) as error:
        raise RuntimeError(f"could not read version marker: {error}") from error

    for statement in module.body:
        if not isinstance(statement, ast.Assign):
            continue
        if not any(isinstance(target, ast.Name) and target.id == "VERSION" for target in statement.targets):
            continue
        if isinstance(statement.value, ast.Constant) and isinstance(statement.value.value, str):
            return statement.value.value

    raise RuntimeError(f"could not find VERSION in {version_marker}")


def fetch_latest_release() -> tuple[str, str] | None:
    """Fetch the latest Cadre CLI version from PyPI.

    Returns (version, release_url) or None if fetch fails.
    Checks PyPI for the canonical pip/pipx distribution version.
    """
    url = "https://pypi.org/pypi/cadre/json"

    try:
        with urllib.request.urlopen(url, timeout=5) as response:
            if response.status != 200:
                return None
            data = json.loads(response.read().decode("utf-8"))
            version = data.get("info", {}).get("version", "")
            release_url = f"https://pypi.org/project/cadre/{version}/"
            return (version, release_url) if version else None
    except (urllib.error.URLError, urllib.error.HTTPError, json.JSONDecodeError, OSError):
        return None


def detect_install_method() -> str:
    """Detect how Cadre was installed: 'pipx', 'pip', or 'source'.

    Returns the detection result, which influences how we recommend updating.
    """
    # Check if running from a source checkout
    # Look for pyproject.toml and bin/cadre at repo root
    if (REPO_ROOT / "pyproject.toml").is_file() and (REPO_ROOT / "bin" / "cadre").is_file():
        return "source"
    # Also check parent directory in case we're in a plugin bundle
    if (REPO_ROOT.parent / "pyproject.toml").is_file() and (REPO_ROOT.parent / "bin" / "cadre").is_file():
        return "source"

    # Try to detect pipx installation by checking if 'cadre' command exists
    # and is managed by pipx
    try:
        result = subprocess.run(
            ["pipx", "list", "--json"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            try:
                data = json.loads(result.stdout)
                if "venvs" in data and "cadre" in data["venvs"]:
                    return "pipx"
            except json.JSONDecodeError:
                pass
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass

    # Default to pip
    return "pip"


def compare_versions(current: str, latest: str) -> int:
    """Compare version strings, handling pre-release versions.

    Returns: -1 if current < latest, 0 if equal, 1 if current > latest
    """
    def parse_version(v: str) -> tuple:
        # Strip pre-release markers (rc, a, b, dev, post, etc.)
        # Examples: 1.2.3rc1 -> 1.2.3, 1.2.3a1 -> 1.2.3
        base = v.split("+")[0]  # Remove local version
        for marker in ("rc", "a", "b", "dev", "post"):
            if marker in base.lower():
                base = base.lower().split(marker)[0]
                break

        try:
            parts = tuple(int(x) for x in base.split("."))
            return parts if parts else (0,)
        except (ValueError, AttributeError):
            return (0,)

    current_parts = parse_version(current)
    latest_parts = parse_version(latest)

    if current_parts < latest_parts:
        return -1
    elif current_parts > latest_parts:
        return 1
    return 0


def prompt_update(current: str, latest: str, release_url: str) -> bool:
    """Prompt the user to confirm an update."""
    print(f"\nCadre update available!")
    print(f"  Current version: {current}")
    print(f"  Latest version:  {latest}")
    print(f"  Release notes:   {release_url}")
    print()

    try:
        response = input("Update Cadre now? (y/n) [n]: ").strip().lower()
        return response == "y"
    except (EOFError, KeyboardInterrupt):
        return False


def update_cadre(install_method: str) -> int:
    """Attempt to update Cadre using the detected installation method.

    Returns 0 on success, 1 on failure.
    """
    if install_method == "source":
        print("cadre: Running from a source checkout. To update, use:")
        print("  git pull origin main")
        print("  ./bin/cadre generate-role-metadata")
        print("  ./bin/cadre generate-plugin --output plugin")
        return 0

    if install_method == "pipx":
        print("Updating via pipx...")
        result = subprocess.run(["pipx", "upgrade", "cadre"])
        if result.returncode == 0:
            print("✓ Cadre updated successfully via pipx")
            print("  Run 'cadre --version' to verify the new version")
        return result.returncode

    # pip
    print("Updating via pip...")
    result = subprocess.run([sys.executable, "-m", "pip", "install", "--upgrade", "cadre"])
    if result.returncode == 0:
        print("✓ Cadre updated successfully via pip")
        print("  Run 'cadre --version' to verify the new version")
    return result.returncode


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        prog="cadre upgrade",
        description="Check for and install Cadre CLI updates",
        add_help=False,  # We'll handle help ourselves to match cadre style
    )
    parser.add_argument(
        "--check",
        action="store_true",
        help="Only check for updates, don't install",
    )
    parser.add_argument(
        "--force",
        action="store_true",
        help="Update without confirmation",
    )
    parser.add_argument(
        "-h",
        "--help",
        action="store_true",
        help="Show this help message",
    )

    try:
        args = parser.parse_args(argv)
    except SystemExit:
        return 2

    if args.help:
        parser.print_help()
        return 0

    try:
        current_version = get_installed_version()
    except RuntimeError as error:
        print(f"cadre upgrade: {error}", file=sys.stderr)
        return 1

    print(f"Current version: {current_version}")
    print("Checking for updates...")

    result = fetch_latest_release()
    if result is None:
        print("cadre upgrade: Could not reach GitHub to check for updates", file=sys.stderr)
        print("  Check your internet connection or try again later")
        return 1

    latest_version, release_url = result

    comparison = compare_versions(current_version, latest_version)

    if comparison == 0:
        print(f"✓ Cadre {current_version} is up to date")
        return 0

    if comparison > 0:
        print(f"✓ Cadre {current_version} is newer than the latest release ({latest_version})")
        return 0

    # comparison < 0: update available
    if args.check:
        print(f"Update available: {current_version} → {latest_version}")
        print(f"Release notes: {release_url}")
        return 0

    install_method = detect_install_method()

    if args.force:
        return update_cadre(install_method)

    if not prompt_update(current_version, latest_version, release_url):
        print("Update cancelled")
        return 0

    return update_cadre(install_method)


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: internal/cli/upgrade.go
package cli

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/deagy/cadre/cli/internal/orchestration"
)

// UpgradeCmd is the `cadre upgrade` command: check for newer versions and apply updates.
// cliReleaseTagPrefix is the tag namespace .github/workflows/release.yml's
// cli-publish job creates for the compiled CLI.
const cliReleaseTagPrefix = "cli-v"

func UpgradeCmd(args []string) int {
	fs := flag.NewFlagSet("cadre upgrade", flag.ContinueOnError)
	fs.SetOutput(os.Stderr)
	checkOnly := fs.Bool("check", false, "only check for updates, don't install")
	force := fs.Bool("force", false, "update without confirmation")
	help := fs.Bool("help", false, "show this help message")
	if err := fs.Parse(args); err != nil {
		return 2
	}
	if *help {
		fmt.Fprintf(os.Stderr, "usage: cadre upgrade [--check|--force|--help]\n")
		fmt.Fprintf(os.Stderr, "\nCheck for and install Cadre CLI updates.\n")
		fmt.Fprintf(os.Stderr, "\n--check   Only check for updates, don't install\n")
		fmt.Fprintf(os.Stderr, "--force   Update without confirmation\n")
		fmt.Fprintf(os.Stderr, "--help    Show this help message\n")
		return 0
	}

	currentVersion, err := getInstalledVersion()
	if err != nil {
		fmt.Fprintf(os.Stderr, "cadre upgrade: %v\n", err)
		return 1
	}
	fmt.Printf("Current version: %s\n", currentVersion)
	fmt.Println("Checking for updates...")

	result, err := fetchLatestRelease()
	if err != nil {
		fmt.Fprintf(os.Stderr, "cadre upgrade: could not reach GitHub to check for a %s* release\n", cliReleaseTagPrefix)
		fmt.Fprintf(os.Stderr, "  check your internet connection or try again later\n")
		return 1
	}
	if result == nil {
		fmt.Fprintf(os.Stderr, "cadre upgrade: could not parse release information\n")
		return 1
	}

	latestVersion := result["version"].(string)
	releaseURL := result["release_url"].(string)

	comparison := compareVersions(currentVersion, latestVersion)

	if comparison == 0 {
		fmt.Printf("✓ Cadre %s is up to date\n", currentVersion)
		return 0
	}

	if comparison > 0 {
		fmt.Printf("✓ Cadre %s is newer than the latest release (%s)\n", currentVersion, latestVersion)
		return 0
	}

	// Update available
	if *checkOnly {
		fmt.Printf("Update available: %s → %s\n", currentVersion, latestVersion)
		fmt.Printf("Release notes: %s\n", releaseURL)
		return 0
	}

	kind, root := detectInstallKind()
	tag, _ := result["tag"].(string)

	// Only the wheel channel is updated in place; every other kind prints an
	// instruction and changes nothing, so there is nothing to confirm.
	if kind != orchestration.InstallKindUnknown {
		return updateCadre(kind, root, tag)
	}

	if *force {
		return updateCadre(kind, root, tag)
	}

	if !promptUpdate(currentVersion, latestVersion, releaseURL) {
		fmt.Println("Update cancelled")
		return 0
	}

	return updateCadre(kind, root, tag)
}

func getInstalledVersion() (string, error) {
	// Use the existing CLIVersion function which locates and parses the
	// VERSION marker from cadre_cli/_version.py (or the vendored wheel layout).
	// First, find the repo root.
	repoRoot, err := findRepoRoot()
	if err != nil {
		return "", err
	}
	return CLIVersion(repoRoot)
}

func findRepoRoot() (string, error) {
	exe, err := os.Executable()
	if err != nil {
		return "", err
	}
	current := filepath.Dir(exe)

	// Search up to 10 levels
	for i := 0; i < 10; i++ {
		if _, err := os.Stat(filepath.Join(current, "bin", "cadre")); err == nil {
			return current, nil
		}
		if _, err := os.Stat(filepath.Join(current, "pyproject.toml")); err == nil {
			return current, nil
		}
		if _, err := os.Stat(filepath.Join(current, "go.mod")); err == nil {
			return current, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}

	// Fallback: assume repo root 3 levels up from this executable
	return filepath.Join(filepath.Dir(exe), "..", "..", ".."), nil
}

// fetchLatestRelease returns the newest `cli-v*` GitHub release.
//
// This used to query PyPI. That was coherent for exactly one of the four
// install kinds: the pip/pipx wheel. A checkout, a `go install` binary and a
// plugin-cache install are none of them updated from PyPI, and a wheel
// publish can lag or fail independently of the binaries. The CLI's own
// release job (.github/workflows/release.yml's cli-publish) tags
// `cli-v<version>` from cadre_cli/_version.py and attaches the per-platform
// binaries there, so that tag is the one source every install kind shares.
//
// /releases/latest is not usable here: this repository also publishes
// `plugin-v*` and kernel tags, and "latest" is whichever was cut most
// recently regardless of prefix.
func fetchLatestRelease() (map[string]any, error) {
	url := "https://api.github.com/repos/deagy/cadre/releases?per_page=30"
	client := &http.Client{Timeout: 5 * time.Second}

	resp, err := client.Get(url)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var releases []map[string]any
	if err := json.Unmarshal(body, &releases); err != nil {
		return nil, err
	}

	for _, release := range releases {
		tag, _ := release["tag_name"].(string)
		if !strings.HasPrefix(tag, cliReleaseTagPrefix) {
			continue
		}
		if draft, _ := release["draft"].(bool); draft {
			continue
		}
		version := strings.TrimPrefix(tag, cliReleaseTagPrefix)
		if version == "" {
			continue
		}
		releaseURL, _ := release["html_url"].(string)
		if releaseURL == "" {
			releaseURL = "https://github.com/deagy/cadre/releases/tag/" + tag
		}
		return map[string]any{"version": version, "release_url": releaseURL, "tag": tag}, nil
	}
	return nil, fmt.Errorf("no %s* release found", cliReleaseTagPrefix)
}

// detectInstallKind reuses the classification cadre doctor already
// performs, rather than re-deriving it here.
//
// The previous implementation looked for a pyproject.toml, then shelled out
// to `pipx list --json`, then defaulted to "pip" -- so a checkout whose
// probe missed was told to run `pip install --upgrade cadre`, installing a
// wheel over the CLI it was already running from a git tree.
// orchestration.ClassifyRunningBinary distinguishes checkout, go-install,
// plugin-cache and unknown, and is covered by its own tests.
func detectInstallKind() (kind string, root string) {
	report := orchestration.GatherDoctorReport("", "")
	return report.InstallKind, report.InstallRoot
}

func compareVersions(current, latest string) int {
	currentParts := parseVersion(current)
	latestParts := parseVersion(latest)

	// Compare tuples
	minLen := len(currentParts)
	if len(latestParts) < minLen {
		minLen = len(latestParts)
	}

	for i := 0; i < minLen; i++ {
		if currentParts[i] < latestParts[i] {
			return -1
		}
		if currentParts[i] > latestParts[i] {
			return 1
		}
	}

	// If all equal so far, check which has more components
	if len(currentParts) < len(latestParts) {
		return -1
	}
	if len(currentParts) > len(latestParts) {
		return 1
	}

	return 0
}

func parseVersion(v string) []int {
	// Extract base version (e.g., "1.2.3" from "1.2.3rc1")
	re := regexp.MustCompile(`^(\d+(?:\.\d+)*)`)
	match := re.FindString(v)
	if match == "" {
		return []int{0}
	}

	parts := strings.Split(match, ".")
	var result []int
	for _, part := range parts {
		if n, err := strconv.Atoi(part); err == nil {
			result = append(result, n)
		}
	}

	// If there's a pre-release suffix, mark it as older by appending -1
	if len(v) > len(match) {
		result = append(result, -1)
	} else {
		result = append(result, 0)
	}

	if len(result) == 0 {
		return []int{0}
	}
	return result
}

func promptUpdate(current, latest, releaseURL string) bool {
	fmt.Printf("\nCadre update available!\n")
	fmt.Printf("  Current version: %s\n", current)
	fmt.Printf("  Latest version:  %s\n", latest)
	fmt.Printf("  Release notes:   %s\n", releaseURL)
	fmt.Println()

	fmt.Print("Update Cadre now? (y/n) [n]: ")
	var response string
	_, _ = fmt.Scanln(&response)
	return strings.ToLower(strings.TrimSpace(response)) == "y"
}

// updateCadre routes on how this binary was installed. Only the wheel
// channel is updated in place; the others print the one correct instruction
// for their kind, because `git pull` in someone's working tree and a
// marketplace refresh are not this command's to perform unprompted.
func updateCadre(kind, root, tag string) int {
	switch kind {
	case orchestration.InstallKindCheckout:
		fmt.Println("cadre: running from a git checkout" + locatedAt(root) + ".")
		fmt.Println("  To update:")
		fmt.Println("    git -C " + displayRoot(root) + " pull --ff-only")
		fmt.Println()
		fmt.Println("  bin/cadre rebuilds the Go binary on the next invocation, so no")
		fmt.Println("  separate build step is needed. If the pull touched roster/,")
		fmt.Println("  .agents/skills/ or AGENTS.md, re-run the regeneration sequence in")
		fmt.Println("  roster/RUNBOOK.md section 17 before committing anything.")
		return 0

	case orchestration.InstallKindGoInstall:
		target := "github.com/deagy/cadre/cmd/cadre@" + tag
		fmt.Println("cadre: installed with `go install`" + locatedAt(root) + ".")
		fmt.Println("  To update:")
		fmt.Println("    go install " + target)
		return 0

	case orchestration.InstallKindPluginCache:
		fmt.Println("cadre: running from a Claude Code plugin cache" + locatedAt(root) + ".")
		fmt.Println("  This copy is managed by the plugin marketplace, not by this command.")
		fmt.Println("  To update, from Claude Code:")
		fmt.Println("    /plugin marketplace update cadre-team")
		fmt.Println("  then reinstall or update the cadre plugin from that marketplace.")
		return 0
	}

	// Unknown: the pip/pipx wheel is the only remaining channel this command
	// can act on, and it is the one PyPI genuinely serves.
	if pipxHasCadre() {
		fmt.Println("Updating via pipx...")
		command := exec.Command("pipx", "upgrade", "cadre")
		command.Stdout = os.Stdout
		command.Stderr = os.Stderr
		if err := command.Run(); err != nil {
			return 1
		}
		fmt.Println("\u2713 Cadre updated successfully via pipx")
		fmt.Println("  Run 'cadre --version' to verify the new version")
		return 0
	}

	fmt.Println("Updating via pip...")
	command := exec.Command("pip", "install", "--upgrade", "cadre")
	command.Stdout = os.Stdout
	command.Stderr = os.Stderr
	if err := command.Run(); err != nil {
		return 1
	}
	fmt.Println("\u2713 Cadre updated successfully via pip")
	fmt.Println("  Run 'cadre --version' to verify the new version")
	return 0
}

// pipxHasCadre reports whether pipx manages a cadre venv.
func pipxHasCadre() bool {
	output, err := exec.Command("pipx", "list", "--json").Output()
	if err != nil {
		return false
	}
	var data map[string]any
	if err := json.Unmarshal(output, &data); err != nil {
		return false
	}
	venvs, ok := data["venvs"].(map[string]any)
	if !ok {
		return false
	}
	_, present := venvs["cadre"]
	return present
}

func locatedAt(root string) string {
	if root == "" {
		return ""
	}
	return " at " + root
}

func displayRoot(root string) string {
	if root == "" {
		return "<checkout>"
	}
	return root
}
